### PR TITLE
AP_Mount: add topotek gimbal driver

### PR DESCRIPTION
**Pull request by rmackay9**
This adds support for the [Topotek gimbals](https://www.topotek.com/) and replaces https://github.com/ArduPilot/ardupilot/pull/26666

This has been heavily reworked to reduced its flash size (now 6660 on CubeOrangePlus) and to make it more robust to errors.

This has been tested on real hardware including confirming that:

- angle control works
- rate control works
- camera take-picture and record-video works
- pictures and videos are recorded to the SD card with the correct date
- pictures include EXIF data including accurate location where the picture was taken

Some of the changes compared to [the initial PR](https://github.com/ArduPilot/ardupilot/pull/26666) are:

- display firmware version to user
- switch automatically to RC control
- simplfy set-gimbal-lock
- replace use of ctime
- use unix line endings
- comment improvements

The only known issues are:

- some use of "int" and "long" (see send_location_info() and gimbal_dist_info_analyse()
- the model version is reported as "Unknown"
- during yaw rate control the gimbal always moves in "earth-frame".  Once it stops rotating (e.g. user centers sticks) the gimbal does correctly maintain either earth-frame or body-frame according to the "lock" vs "follow" state.  The DJI RS2/RS3 gimbals also act like this so it's not a blocker

@laozhou-fujian this builds on your PR and the single commit is now Co-authored by both of us.  Can you confirm you are OK with these changes?

Size comparison
image

Wiki PR https://github.com/ArduPilot/ardupilot_wiki/pull/6092

### CODE REVIEW

**Review by Hwurzburg — DISMISSED** [review]
needs MNT_TYPE metadata addition

File: libraries/AP_Mount/AP_Mount_Topotek.cpp [hunk 1]
```diff
@@ -0,0 +1,995 @@
...
+    }
+    else if (('a' <= data) && ('z' >= data)) {
+        return (data - 'a' + 10);
+    }
+    else if (('A' <= data) && ('Z' >= data)) {
+        return (data - 'A' + 10);
+    }
+    return 0;
+}
+
```
Comment (peterbarker) [inline libraries/AP_Mount/AP_Mount_Topotek.cpp:963]: These methods are all available elsewhere in ArduPilot

File: libraries/AP_Mount/AP_Mount_Topotek.cpp [hunk 2]
```diff
@@ -0,0 +1,995 @@
...
+int8_t AP_Mount_Topotek::_next_pip_mode[15] = "#TPUD2wPIP0A";                  // picture-in-picture command
+int8_t AP_Mount_Topotek::_get_gimbal_attitude[15] = "#TPUG2wGAA01";            // get gimbal attitude command
+int8_t AP_Mount_Topotek::_get_gimbal_sdcard_info[15] = "#TPUD2rSDC00";         // get gimbal memory card information command
+int8_t AP_Mount_Topotek::_get_gimbal_track_status[15] = "#TPUD2rTRC00";        // get gimbal tracking status command
+int8_t AP_Mount_Topotek::_get_gimbal_basic_info[15] = "#tpUD2rVSN00";          // get gimbal basic information command
+int8_t AP_Mount_Topotek::_set_yaw_angle_cmd[20] = "#tpUG6wGIY";                // set gimbal yaw angle command
+int8_t AP_Mount_Topotek::_set_pitch_angle_cmd[20] = "#tpUG6wGIP";              // set gimbal pitch angle command
+int8_t AP_Mount_Topotek::_set_roll_angle_cmd[20] = "#tpUG6wGIR";               // set gimbal roll angle command
+int8_t AP_Mount_Topotek::_set_yaw_pitch_roll_speed_cmd[20] = "#tpUG6wYPR";     // set the speed of gimbal yaw, pitch and roll command
+int8_t AP_Mount_Topotek::_set_gimbal_time[45] = "#tpUDCwUTC";                  // set the gimbal time command
```
Comment (tridge) [inline libraries/AP_Mount/AP_Mount_Topotek.cpp:45]: why 45 here for a 10 character string?

File: libraries/AP_Mount/AP_Mount_Topotek.h [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    float _zoom_value;                                          // current zoom value
+    uint32_t _firmware_ver = 0;                                 // firmware version
+    Vector3f _current_angle_rad;                                // current angles in radians received from gimbal (x=roll, y=pitch, z=yaw)
+    uint32_t _last_current_angle_ms = 0;                        // system time (in milliseconds) that angle information received from the gimbal
+    uint32_t _last_req_current_info_ms = 0;                     // system time that this driver last requested current gimbal infomation
+    uint8_t _last_req_step;                                     // 10hz request loop step (different requests are sent at various steps)
+    uint8_t _stop_order_count = 0;                              // number of stop commands sent since target rates became zero
+    float _measure_dist_m = -1.0f;                              // latest rangefinder distance (in meters)
+    int8_t _msg_buff[AP_MOUNT_TOPOTEK_PACKETLEN_MAX] {};        // buffer holding bytes from latest packet received.  only used to calculate crc
```
Comment (tridge) [inline libraries/AP_Mount/AP_Mount_Topotek.h:197]: why init this? just wastes flash

File: libraries/AP_Mount/AP_Mount_Topotek.h [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    // members
+    bool _initialised;                                          // true once the serial port has been initialised
+    bool _recording;                                            // recording status (received from gimbal)
+    bool _is_tracking;                                          // whether to enable the tracking state
+    bool _sdcard_status;                                        // memory card status (received from gimbal)
+    bool _last_lock;                                            // last lock mode sent to gimbal
+    bool _got_gimbal_basic_info;                                // whether the gimbal information is obtained
+    bool _last_zoom_stop = false;                               // record the latest zoom stopped state.
+    bool _last_focus_stop = false;                              // record the latest focus stopped state.
+    uint8_t _sent_time_count = 0;                               // count of current time messages sent to gimbal
```
Comment (tridge) [inline libraries/AP_Mount/AP_Mount_Topotek.h:186]: why init?

File: libraries/AP_Mount/AP_Mount_Topotek.h [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    int8_t _model_name[11] = "";                                // gimbal type
+    ZoomType _zoom_type;                                        // current zoom type
+    float _zoom_value;                                          // current zoom value
+    uint32_t _firmware_ver = 0;                                 // firmware version
```
Comment (tridge) [inline libraries/AP_Mount/AP_Mount_Topotek.h:190]: wasted flash init

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Mount/AP_Mount_Topotek.cpp [hunk 6]
```diff
@@ -0,0 +1,995 @@
...
+    if (longitude < 0) {
+        longitude = -longitude;
+    }
+
+    // get the degree part of the latitude
+    int lat_deg = (int)latitude;
+
+    // get the minute part of the latitude
+    double lat_min = (latitude - lat_deg) * 60.0;
+    sprintf((char*)_set_gimbal_lat + 10, "%c%02d%07.4f", lat > 0 ? 'N':'S', lat_deg, lat_min);
```
Comment (peterbarker) [inline libraries/AP_Mount/AP_Mount_Topotek.cpp:753]: `_set_gimbal_lat` can just be a local

File: libraries/AP_Mount/AP_Mount_Topotek.cpp [hunk 7]
```diff
@@ -0,0 +1,995 @@
...
+    Location loc;
+    int32_t alt_amsl_cm = 0;
+    if (!AP::ahrs().get_location(loc) || !loc.get_alt_cm(Location::AltFrame::ABSOLUTE, alt_amsl_cm)) {
+        return false;
+    }
+    int32_t lat = loc.lat;
+    int32_t lng = loc.lng;
+
+    double latitude = lat / 10000000.0;
+    double longitude = lng / 10000000.0;
```
Comment (peterbarker) [inline libraries/AP_Mount/AP_Mount_Topotek.cpp:738]: Don't need to have those

File: libraries/AP_Mount/AP_Mount_Topotek.cpp [hunk 8]
```diff
@@ -0,0 +1,995 @@
...
+
+    _set_yaw_angle_cmd[10] = hex2char((yaw_angle_cd >> 12) & 0x0F);
+    _set_yaw_angle_cmd[11] = hex2char((yaw_angle_cd >> 8) & 0x0F);
+    _set_yaw_angle_cmd[12] = hex2char((yaw_angle_cd >> 4) & 0x0F);
+    _set_yaw_angle_cmd[13] = hex2char((yaw_angle_cd) & 0x0F);
+
+    _set_yaw_angle_cmd[14] = hex2char((speed >> 4) & 0x0F);
+    _set_yaw_angle_cmd[15] = hex2char((speed) & 0x0F);
+
+    send_packet(_set_yaw_angle_cmd, 16);
```
Comment (tridge) [inline libraries/AP_Mount/AP_Mount_Topotek.cpp:640]: maybe send_packet(const char *header, const char *fmt, ...), calls snprintf()
send_packet("#tpUG6wGIY", "%04x%02x", yaw_angle_cd, speed);

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Mount/AP_Mount_Topotek.cpp (hunk 8 above)
Comment (tpwrules) [reply libraries/AP_Mount/AP_Mount_Topotek.cpp:640]: The header could be a part of the fmt too.

**Comment by rmackay9** [comment]
We discussed this on the dev call and the major concern was that the driver consumes far too much flash (about 9K).  Various devs have given suggestions on how it could be made smaller so I will attempt to rework the driver based on this advice.

**Comment by rmackay9** [comment]
@laozhou-fujian,

This came up on the dev call and Tridge and I agreed that we should help make this driver smaller.  I've got a 2 week vacation coming up soon but hopefully we can help.

**Comment by laozhou-fujian** [comment]
> @laozhou-fujian,
> 
> This came up on the dev call and Tridge and I agreed that we should help make this driver smaller. I've got a 2 week vacation coming up soon but hopefully we can help.

 Thank you for your help. I will try my best to complete it this week. Thank you.

**Comment by laozhou-fujian** [comment]
> @laozhou-fujian,
> 
> This came up on the dev call and Tridge and I agreed that we should help make this driver smaller. I've got a 2 week vacation coming up soon but hopefully we can help.

Sorry, I've been very busy recently. I will finish it as soon as possible

**Comment by rmackay9** [comment]
@laozhou-fujian,

Just FYI, I've started shrinking the driver and I expect I will complete it within the next week.  No need to stress, I just wanted to make sure we don't waste effort by both doing the same thing.

**Comment by laozhou-fujian** [comment]
> @laozhou-fujian,
> 
> Just FYI, I've started shrinking the driver and I expect I will complete it within the next week. No need to stress, I just wanted to make sure we don't waste effort by both doing the same thing.

Thank you for your advice. I will revise it right away. I feel very sorry for having to deal with a lot of urgent things recently

File: libraries/AP_Mount/AP_Mount_Topotek.cpp (hunk 2 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount_Topotek.cpp]: thanks, these lines have been removed

File: libraries/AP_Mount/AP_Mount_Topotek.h (hunk 4 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount_Topotek.h]: thanks.  all initialisation of members to zero have been removed.

File: libraries/AP_Mount/AP_Mount_Topotek.h (hunk 5 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount_Topotek.h]: gone now, thanks

File: libraries/AP_Mount/AP_Mount_Topotek.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount_Topotek.cpp]: Thanks @peterbarker, I've replaced char-to-number with AP_Common's char-to-hex.  I could not find an equivalent to hex2char though.  I initially thought that AP_Common's hex_to_uint8() was the same but it's actually for a completely different purpose as far as I can tell.

File: libraries/AP_Mount/AP_Mount_Topotek.h (hunk 3 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount_Topotek.h]: init removed, thanks

File: libraries/AP_Mount/AP_Mount_Topotek.cpp (hunk 6 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount_Topotek.cpp]: Txs, I've removed _set_gimbal_lat and all other similar arrays.

File: libraries/AP_Mount/AP_Mount_Topotek.cpp (hunk 7 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount_Topotek.cpp]: Txs, I've made send_location_info() more concise although the two lines you mention are still mostly there.  please tell me if you're not OK with the new form.

File: libraries/AP_Mount/AP_Mount_Topotek.cpp (hunk 8 above)
Comment (rmackay9) [reply libraries/AP_Mount/AP_Mount_Topotek.cpp]: The send_packet() method has been replaced by send_variablelen_packet() which uses less flash and better matches our other drivers.

**Review by tridge — APPROVED** [review]

**Pull request merged**